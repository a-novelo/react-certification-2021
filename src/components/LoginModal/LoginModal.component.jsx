import React, { useState } from 'react';
import { useHistory } from 'react-router';
import { faExclamationCircle } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  LoginModalWrapper,
  LoginContainer,
  FailedLoginContainer,
  LoginForm,
  LoginField,
  LoginLabel,
  LoginFieldTitle,
  LoginFieldInput,
  LoginButtonsContainer,
  LoginButton,
} from './LoginModal.styled';
import { useAuth } from '../../providers/Auth';

function LoginModal(props) {
  const { showModal, handleClose } = props;
  const { login } = useAuth();
  const history = useHistory();

  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const [showFailedLogin, setShowFailedLogin] = useState(false);

  function authenticate(event) {
    event.preventDefault();
    const user = login(username, password);
    if (user === undefined) {
      setShowFailedLogin(true);
    } else {
      handleClose(false);
      history.push('/');
    }
  }

  return (
    <LoginModalWrapper show={showModal}>
      <LoginContainer>
        Login
        <FailedLoginContainer show={showFailedLogin}>
          <FontAwesomeIcon icon={faExclamationCircle} />
          Invalid username and/or password
        </FailedLoginContainer>
        <LoginForm onSubmit={authenticate}>
          <LoginField>
            <LoginLabel htmlFor="username">
              <LoginFieldTitle>username </LoginFieldTitle>
              <LoginFieldInput
                required
                type="text"
                id="username"
                value={username}
                onChange={(e) => {
                  setUsername(e.target.value);
                }}
              />
            </LoginLabel>
          </LoginField>
          <LoginField>
            <LoginLabel htmlFor="password">
              <LoginFieldTitle>password </LoginFieldTitle>
              <LoginFieldInput
                required
                type="password"
                id="password"
                value={password}
                onChange={(e) => {
                  setPassword(e.target.value);
                }}
              />
            </LoginLabel>
          </LoginField>
          <LoginButtonsContainer>
            <LoginButton
              type="button"
              onClick={() => {
                handleClose(false);
              }}
            >
              CANCEL
            </LoginButton>
            <LoginButton type="submit">LOGIN</LoginButton>
          </LoginButtonsContainer>
        </LoginForm>
      </LoginContainer>
    </LoginModalWrapper>
  );
}

export default LoginModal;
