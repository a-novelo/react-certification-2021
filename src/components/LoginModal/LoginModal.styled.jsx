import styled from 'styled-components';

const LoginModalWrapper = styled.div`
  backdrop-filter: blur(8px);
  background-color: rgba(0, 0, 0, 0.4);
  display: ${(props) => (props.show ? 'block' : 'none')};
  font-size: 15px;
  height: 100%;
  left: 0;
  margin: 0 auto;
  overflow: auto;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 999;
`;

const LoginContainer = styled.div`
  background-color: white;
  border-radius: 25px;
  border: none;
  color: black;
  margin: 5% auto;
  padding: 20px;
  width: 60%;
`;

const FailedLoginContainer = styled.div`
  background-color: #ffe4ec;
  color: rgb(102, 9, 27);
  display: ${(props) => (props.show ? 'block' : 'none')};
`;

const LoginForm = styled.form`
  align-items: center;
  display: flex;
  flex-direction: column;
`;

const LoginField = styled.div`
  display: flex;
  flex-direction: column;
  margin-bottom: 1rem;
  width: 100%;
`;

const LoginLabel = styled.label``;

const LoginFieldTitle = styled.strong`
  display: block;
  font-weight: 700;
  margin-bottom: 0.4rem;
  text-transform: capitalize;
`;

const LoginFieldInput = styled.input`
  background-color: rgba(0, 0, 0, 0.1);
  border-radius: 3px;
  border: 1px solid white;
  font-size: 1.2rem;
  padding: 0.4rem 0.6rem;
  width: 100%;
`;

const LoginButtonsContainer = styled.div``;

const LoginButton = styled.button`
  border-radius: 3px;
  border: none;
  font-size: 1.2rem;
  margin: 1rem;
  padding: 0.4rem 0.6rem;
`;

export {
  LoginModalWrapper,
  LoginContainer,
  FailedLoginContainer,
  LoginForm,
  LoginField,
  LoginLabel,
  LoginFieldTitle,
  LoginFieldInput,
  LoginButtonsContainer,
  LoginButton,
};
