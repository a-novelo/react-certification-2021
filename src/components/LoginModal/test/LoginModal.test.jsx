import React from 'react';
import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';
import LoginModal from '../LoginModal.component';
import AuthProvider from '../../../providers/Auth';

describe('#LoginModal', () => {
  it('displays correctly', () => {
    const tree = renderer
      .create(
        <AuthProvider>
          <LoginModal />
        </AuthProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('enters a username and password', () => {
    render(
      <AuthProvider>
        <LoginModal />
      </AuthProvider>
    );
    const inputUsername = screen.getByLabelText('username');
    fireEvent.change(inputUsername, { target: { value: 'Wizeline' } });
    const inputPassword = screen.getByLabelText('password');
    fireEvent.change(inputPassword, { target: { value: 'mypassword' } });
  });

  it('submit username and password', () => {
    render(
      <AuthProvider>
        <LoginModal />
      </AuthProvider>
    );
    const inputUsername = screen.getByLabelText('username');
    fireEvent.change(inputUsername, { target: { value: 'Wizeline' } });
    const inputPassword = screen.getByLabelText('password');
    fireEvent.change(inputPassword, { target: { value: 'wrongpassword' } });
    const loginButton = screen.getByText('LOGIN');
    fireEvent(
      loginButton,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    );
  });

  it('closes modal', () => {
    render(
      <AuthProvider>
        <LoginModal handleClose={() => {}} />
      </AuthProvider>
    );
    const cancelButton = screen.getByText('CANCEL');
    fireEvent(
      cancelButton,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    );
  });
});
