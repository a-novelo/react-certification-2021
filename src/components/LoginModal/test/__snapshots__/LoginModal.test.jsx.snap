// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#LoginModal displays correctly 1`] = `
<div
  className="sc-bdfBwQ laJPru"
>
  <div
    className="sc-gsTCUz hmDnLl"
  >
    Login
    <div
      className="sc-dlfnbm hBmNZb"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-exclamation-circle fa-w-16 "
        data-icon="exclamation-circle"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
      Invalid username and/or password
    </div>
    <form
      className="sc-hKgILt kmZCdh"
      onSubmit={[Function]}
    >
      <div
        className="sc-eCssSg hiLOPZ"
      >
        <label
          className="sc-jSgupP"
          htmlFor="username"
        >
          <strong
            className="sc-gKsewC kYYzDq"
          >
            username 
          </strong>
          <input
            className="sc-iBPRYJ dHHbIC"
            id="username"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
        </label>
      </div>
      <div
        className="sc-eCssSg hiLOPZ"
      >
        <label
          className="sc-jSgupP"
          htmlFor="password"
        >
          <strong
            className="sc-gKsewC kYYzDq"
          >
            password 
          </strong>
          <input
            className="sc-iBPRYJ dHHbIC"
            id="password"
            onChange={[Function]}
            required={true}
            type="password"
            value=""
          />
        </label>
      </div>
      <div
        className="sc-fubCfw"
      >
        <button
          className="sc-pFZIQ kKCbVj"
          onClick={[Function]}
          type="button"
        >
          CANCEL
        </button>
        <button
          className="sc-pFZIQ kKCbVj"
          type="submit"
        >
          LOGIN
        </button>
      </div>
    </form>
  </div>
</div>
`;
